Update sidebar

KTibow opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

Updates the sidebar to
- use raw links instead of paper-listbox, fixes some ugly apperance stuff and ctrl+clicking
- use m3 styling
- refactor so it isn't a megacomponent

## Type of change

<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
NA
## Additional information

<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist

<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io

KTibow committed:
initial sidebar refactoring work, todo: sorting

KTibow committed:
add editing and fix stuff

KTibow committed:
rename some stuff and add external config

KTibow commented:
| Original | M3 Example | Implementation |
|--------|--------|--------|
| ![image](https://user-images.githubusercontent.com/10727862/229361041-cf1ba506-36c0-418c-b95d-cad3161a247a.png) | ![image](https://user-images.githubusercontent.com/10727862/229361220-7a9d82d7-0a99-4bc3-81c6-149dce33cd82.png) | ![image](https://user-images.githubusercontent.com/10727862/229361093-9971bbaf-2469-484c-9790-f12e87299a08.png) |
| ![image](https://user-images.githubusercontent.com/10727862/229361106-b0e67f31-5f8c-4d56-83f1-a56031af09c8.png) | ![image](https://user-images.githubusercontent.com/10727862/229361149-0bf57f81-583d-419e-9439-220ec148d837.png) | ![image](https://user-images.githubusercontent.com/10727862/229361171-b888eb75-8905-4f55-b963-64ca57064f05.png) |

bramkragten commented on src/components/ha-sidebar.ts:
```diff
@@ -198,934 +41,88 @@ class HaSidebar extends SubscribeMixin(LitElement) {
+  private _mouseOverItem(ev: { target: HTMLElement }) {
+    const tooltipRoot =
+      this.shadowRoot?.querySelector("ha-sidebar-tooltip")?.shadowRoot;
+    const tooltip = tooltipRoot?.firstElementChild as HTMLElement;
```
Why not just use a `slot`?

KTibow replied:
i don't think a slot would make things easier from what i understand, the tooltip would still be in a shadow root, and you would require even more logic to traverse the slot

bramkragten commented:
> use raw links instead of paper-listbox, fixes some ugly apperance stuff and ctrl+clicking

With the list element removed, we now no longer have keyboard navigation, like typeahead and simple arrow handling. Also is it now missing a lot of a11y feats.

KTibow commented:
re: a11y, after checking on my machine:
- i do need to add `tabindex="-1"` to one of the items and do that `role="option"` thing
- what is "typeahead and simple arrow handling"?
- keyboard navigation still works with shift-tab/tab and enter

KTibow committed:
fix focusing the list of items

KTibow committed:
fix rtl (the old sidebar had a rtl bug, fun fact)

KTibow committed:
Merge branch 'dev' into dev

LasseRosenow commented:
Maybe better to just wait for the [@material/web](https://github.com/material-components/material-web) drawer component?

bramkragten commented:
> re: a11y, after checking on my machine:
> 
> * what is "typeahead and simple arrow handling"?
> * keyboard navigation still works with shift-tab/tab and enter

Press a letter when focus is on the old sidebar list, it will focus the item starting with the letter you pressed

KTibow commented:
@LasseRosenow there is a nav drawer component but that just wraps whatever content you want to put inside of it - could be they'll add more later as it's marked as wip
@bramkragten what a11y specs are you referring to

bramkragten commented:
Basic a11y features for lists <https://w3c.github.io/aria/#list>

KTibow commented:
Well I'm not sure how to move forward on this, as using paper-listbox would reintroduce the bugs like ctrl+clicking behavior. Should I make a custom implementation to handle those keys?

KTibow committed:
Merge branch 'dev' into dev

bramkragten commented:
> Well I'm not sure how to move forward on this, as using paper-listbox would reintroduce the bugs like ctrl+clicking behavior. Should I make a custom implementation to handle those keys?

You can use `mwc-list` but that will not work with anchor links, so open in new tab etc will no longer work.

KTibow committed:
Fix accessibility and other bugs
- fix currentpanel logic
- disable native outline, indicate focus on active page
- prevent focus on container
- space works to activate an item
- typing and up/down works to focus an item

KTibow commented:
I believe this PR should be ready for comment now. In my testing I didn't find any bugs, and everything from the ARIA listbox guidelines (if it was in there previously) is now working.

KTibow committed:
Merge branch 'dev' into dev

KTibow committed:
use @storage

silamon commented:
Too bad this didn't got a review yet, I've had a go and tested your changes. I didn't look at the code yet. 

The sidebar is so much smoother by removing that paper-listbox. It feels like the dashboard also loads faster. 
I hope you don't mind if I share some things that didn't work for me: 
* Reordering the dashboards: I can go in edit mode, but dragging dashboards to other places doesn't seem to move it. 
* The scrollbar takes a lot of place as can be seen in the screenshot
![image](https://github.com/home-assistant/frontend/assets/32477463/341565ee-77ef-4b25-955a-f93c56a0e0d2)

KTibow commented:
i'll see if i can reproduce that

KTibow committed:
improve scrolling

silamon commented on src/components/ha-sidebar-panels.ts:
```diff
@@ -0,0 +1,344 @@
+
+@customElement("ha-sidebar-panels")
+class HaSidebarPanels extends LitElement {
+  static shadowRootOptions = {
+    ...LitElement.shadowRootOptions,
+    delegatesFocus: true,
+  };
```
This is the reason the ordering of dashboards no longer work.

KTibow replied:
Can't reproduce the issue with reordering dashboards
Can you send a screen recording and other relevant info

silamon replied:
So, I can reproduce this on both Microsoft Edge and Google Chrome latest version. On Mozilla Firefox it works good. I'll see if I can create a screen recording.

KTibow replied:
Can't reproduce w/ Chromium 117.0.5938.132 or Chrome Mobile 118.0.5993.80
Does it work if you clear your localstorage by chance

silamon replied:
Too bad, removing localstorage doesn't solve it. It works if the shadowRootOptions are removed or delegatesFocus is set to false. Let's wait until somebody other reviews.

bramkragten commented:
Major UI changes like these will have to be approved by our design department, CC @matthiasdebaat

KTibow commented:
I see that there was a "needs design preview" label added. Assuming that refers to https://design.home-assistant.io/, I have some questions about how to implement one:
- How exactly do I display the sidebar? Am I supposed to display the whole sidebar or individual panels? If I need to display the whole sidebar:
  - What kinds of states do I need to display? Do I need to show what it looks like with devtools, on mobile, etc? If so, should I just display all the different kinds of sidebars, or provide switches that can be used?
  - Where can I learn about how to set up a fake hass?
- As someone who isn't as familiar with the frontend as someone in the core team might be, is there a specific contributing guide for how to create a design preview?
- Also, is there any specific reason a design preview was requested? The previous sidebar isn't there. Is this a norm for new UI components? Does this make it easier to review? As you can tell I'm not sure if I know enough to create a design preview.

matthiasdebaat commented:
Thank you for helping us update our components to Material 3. From a UX point of view, this PR should only update the visual design, with a few exceptions:
- M3 design language is bolder than 2, which takes up more space. The disadvantage is that fewer menu items are displayed if we follow these guidelines. Our users have a lot of menu items in the sidebar, showing less would not help our users. In this case, we need to keep the height of the menu items the same as we currently have.
- While I like the M3 approach of showing a number instead of a badge, we need to keep the collapsed and expanded menu consistent. Let's keep our badge in both states.

Curious how the edit "wiggle" mode looks. Great if we can add the sidebar to our design website.

> I see that there was a "needs design preview" label added. Assuming that refers to https://design.home-assistant.io/, I have some questions about how to implement one:

@bramkragten can help you with this.

KTibow commented:
> In this case, we need to keep the height of the menu items the same as we currently have.

Crazy idea but what do you think about allowing the menu items to compress if there's too many? Of course if there's too many they would just max out at the previous UI height and scroll the menu.

> Let's keep our badge in both states.

[TODO: check if the badge still breaks at 10+ numbers]

> Curious how the edit "wiggle" mode looks

Edit works as usual, except I didn't implement the wiggling yet.


To summarize, soon I'll implement
- some sort of change to the item height (once I hear what other people think about a dynamic item height)
- using the badge in more cases
- wiggling animation
- larger touch targets
- fixing merge conflicts

matthiasdebaat commented:
> Crazy idea but what do you think about allowing the menu items to compress if there's too many? Of course if there's too many they would just max out at the previous UI height and scroll the menu.

Interesting idea, have to see it if we should implement it. Couple of downsides is that it can get bouncy or have unexpected behaviour for our users.

KTibow committed:
Merge branch 'dev' into dev

KTibow committed:
fix most stuff (TODO: wiggling)

karwosts commented:
`needs design preview` is supposed to trigger a github task to automatically build a new copy of design.home-assistant.io with the changes in this PR. However I believe the action can not work for PRs that originate from forks, so it won't work for this case.

KTibow committed:
somewhat improve sorting

KTibow commented:
Looks like the "deploy preview" workflow is asking for the user to open a browser to sign in, so it failed to deploy. That's also why it took so long - it timed out.

KTibow committed:
Merge branch 'home-assistant:dev' into dev

KTibow commented:
Hey, I'd just like to recap that with `3f1fe22` and `586e12f`, I added/fixed these things:
- A reduced item height
- Using the full badge when expanded
- A wiggling animation when in edit mode
- Full touch targets

If the code looks all good, this PR is ready to merge.

KTibow committed:
Merge remote-tracking branch 'upstream/dev' into dev

KTibow committed:
fix a11y

KTibow committed:
Merge branch 'dev' into dev

KTibow committed:
Remove useless slots

KTibow committed:
Fix selected

KTibow committed:
Fix RTL

KTibow commented:
Should be resolved on the latest commit

KTibow committed:
Merge branch 'dev' into dev

KTibow committed:
Merge remote-tracking branch 'upstream/dev' into dev

KTibow committed:
format code

silamon commented:
@bramkragten can this get a review somehow or is there anything that still needs to be done before review?

KTibow committed:
Merge remote-tracking branch 'upstream/dev' into dev

KTibow committed:
fix formatting

bramkragten commented:
I don't like the text color and especially the icon color of the sidebar, is there a reason these are now so dark?
Also the text "Home Assistant" on top of the expanded sidebar looks very small?
I also miss things like the ripple effect in your implementation, but that might be ok.

KTibow committed:
address review

KTibow commented:
> I don't like the text color and especially the icon color of the sidebar, is there a reason these are now so dark?

In the dark theme or light theme? Thing is that M3 calls for `on-surface-variant`, but we only have `primary-text-color` and `secondary-text-color` (a much less visible color). While I could decrease the opacity (to 90%) to match `on-surface-variant`, since I use rgba and var in my code and you can't multiply rgba, this wouldn't work.

> Also the text "Home Assistant" on top of the expanded sidebar looks very small?

Yup, this is in compliance with Material 3. The title should be `title-small`, which is 14pt. Technically speaking it should be even smaller as Home Assistant uses a text size that is 7/8 the normal size.

> I also miss things like the ripple effect in your implementation, but that might be ok.

The previous implementation didn't have ripples, and I'm not exactly sure which ripples to use.

bramkragten commented on src/components/ha-sidebar-panel-config.ts:
```diff
@@ -30,17 +31,19 @@ class HaSidebarPanelConfig extends LitElement {
-
   protected render() {
     const notices = this._updatesCount + this._issuesCount;
+    const keydown = createKeydown((e) =>
+      (e.currentTarget as HTMLElement).click()
+    );
+    const keyup = createKeyup((e) => (e.currentTarget as HTMLElement).click());
```
This is still the same, as every time on render a new function is created and a new event listener is added. This should just be a class function.

KTibow committed:
address review (make keydown/keyup static)

codyc1515 commented:
Current or can be closed?

KTibow commented:
Let me fix the conflicts (yet again..)

KTibow commented:
Actually there have been too many changes for me to want to bother with fixing this yet again. I'm just sad that this never got a chance to get merged.
